arangod/V8Server/v8-user-functions.cpp

```diff
@@ -0,0 +1,153 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2016 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Wilfried Goesgens
+////////////////////////////////////////////////////////////////////////////////
+
+#include "v8-user-functions.h"
+
+//#include "Basics/Result.h"
+
+#include "v8-vocbaseprivate.h"
+#include <v8.h>
+#include "V8/v8-utils.h"
+#include "V8/v8-vpack.h"
+#include "V8/v8-conv.h"
+
+#include "VocBase/Methods/AqlUserFunctions.h"
+
+using namespace arangodb;
+
+static void JS_UnregisterAQLUserFunction(v8::FunctionCallbackInfo<v8::Value> const& args) {
+  TRI_V8_TRY_CATCH_BEGIN(isolate);
+  v8::HandleScope scope(isolate);
+
+  if (args.Length() != 1) {
+    TRI_V8_THROW_EXCEPTION_USAGE("UNREGISTER_AQL_USER_FUNCTION(<name>)");
+  }
+
+  TRI_vocbase_t* vocbase = GetContextVocBase(isolate);
+
+  std::string functionName = TRI_ObjectToString(args[0]);
+
+  Result rv = unregisterUserFunction(vocbase, functionName);
+
+  if (rv.fail()) {
+    TRI_V8_THROW_EXCEPTION(rv);
+  }
+  TRI_V8_TRY_CATCH_END;
+}
+
+static void JS_UnregisterAQLUserFunctionsGroup(v8::FunctionCallbackInfo<v8::Value> const& args) {
+  TRI_V8_TRY_CATCH_BEGIN(isolate);
+  v8::HandleScope scope(isolate);
+
+  if (args.Length() != 1) {
+    TRI_V8_THROW_EXCEPTION_USAGE("UNREGISTER_AQL_USER_FUNCTION_GROUP(<group string>)");
+  }
+
+  int deleteCount;
+  TRI_vocbase_t* vocbase = GetContextVocBase(isolate);
+
+  std::string functionFilterPrefix = TRI_ObjectToString(args[0]);
+
+  Result rv = unregisterUserFunctionsGroup(vocbase, functionFilterPrefix, deleteCount);
+
+  if (rv.fail()) {
+    TRI_V8_THROW_EXCEPTION(rv);
+  }
+  TRI_V8_RETURN(v8::Number::New(isolate, deleteCount));
+  TRI_V8_TRY_CATCH_END;
+}
+
+static void JS_RegisterAqlUserFunction(v8::FunctionCallbackInfo<v8::Value> const& args) {
+  TRI_V8_TRY_CATCH_BEGIN(isolate);
+  v8::HandleScope scope(isolate);
+
+  if (args.Length() != 1) {
+    TRI_V8_THROW_EXCEPTION_USAGE("REGISTER_AQL_USER_FUNCTION(<name>, <functionbody> [, <isDeterministic>])");
+  }
+
+  TRI_normalize_V8_Obj(args, args[0]);
+
+  TRI_vocbase_t* vocbase = GetContextVocBase(isolate);
+
+  VPackBuilder builder;
+  TRI_V8ToVPackSimple(isolate, builder, args[0]);
```

Comment:
return value of function call is not check here.
Instead of `TRI_V8ToVPackSimple`, one should definitely call `TRI_V8ToVPack` here.
The precondition for `TRI_V8ToVPackSimple` is according to its docs:

> this function assumes that the V8 object does not contain any cycles and does not contain types such as Function, Date or RegExp

Not sure if we can guarantee this, so we should rather use TRI_V8ToVPack

Reply:
Hm, the .js wrapper does the function->string conversion, shouldn't that be good enough then?

Reply:
who is keeping anyone from calling the C++ function directly?